pkg/graveler/graveler_v2_test.go

```diff
@@ -547,3 +550,132 @@ func TestGravelerCommit_v2(t *testing.T) {
+			numBranch:  1000,
+			numRecords: 2000,
+			offset:     2053,
+		},
+		{
+			name:       "WithMark after max",
+			numBranch:  1000,
+			numRecords: 2000,
+			offset:     graveler.GCMaxUncommittedObjects + 1234,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			test := createTestScenario(t, tt.numBranch, tt.numRecords, tt.offset)
+			mark := graveler.GCUncommittedMark{
+				BranchID: "",
+				Key:      []byte(""),
+			}
+			if tt.offset != 0 {
+				branchID := fmt.Sprintf("branch%04d", tt.offset/tt.numRecords)
+				mark.BranchID = graveler.BranchID(branchID)
+				mark.Key = graveler.Key(fmt.Sprintf("%s_record%04d", branchID, tt.offset%tt.numRecords))
+			}
+
+			list, NewMark, err := test.sut.GetGarbageCollectionUncommitted(ctx, repository, mark)
+			require.NoError(t, err)
+			totalRecords := tt.numRecords*tt.numBranch - tt.offset
+			totalExpected := math.Min(float64(totalRecords), graveler.GCMaxUncommittedObjects)
+			require.Equal(t, int(totalExpected), len(list))
+			if totalRecords <= graveler.GCMaxUncommittedObjects {
+				require.Nil(t, NewMark)
+			} else {
+				branch := (graveler.GCMaxUncommittedObjects + tt.offset) / tt.numRecords
+				recordID := (graveler.GCMaxUncommittedObjects + tt.offset) % tt.numRecords
+				branchID := graveler.BranchID(fmt.Sprintf("branch%04d", branch))
+				require.Equal(t, branchID, NewMark.BranchID)
+				key := fmt.Sprintf("%s_record%04d", branchID, recordID)
+				require.Equal(t, key, string(NewMark.Key))
+			}
+			verifyData(t, list, tt.numRecords, tt.offset)
+		})
+	}
+}
+
+func createTestScenario(t *testing.T, numBranches, numRecords, offset int) *gravelerTest {
+	t.Helper()
+	test := initGravelerTest(t)
+	records := make([][]*graveler.ValueRecord, numBranches)
+	var branches []*graveler.BranchRecord
+	for i := 0; i < numBranches; i++ {
+		branchID := graveler.BranchID(fmt.Sprintf("branch%04d", i))
+		token := graveler.StagingToken(fmt.Sprintf("%s_st%04d", branchID, i))
+		branches = append(branches, &graveler.BranchRecord{BranchID: branchID, Branch: &graveler.Branch{StagingToken: token}})
+
+		records[i] = make([]*graveler.ValueRecord, numRecords)
+		for j := 0; j < numRecords; j++ {
+			records[i][j] = &graveler.ValueRecord{
+				Key:   []byte(fmt.Sprintf("%s_record%04d", branchID, j)),
+				Value: nil,
+			}
+		}
+	}
+	test.refManager.EXPECT().ListBranches(gomock.Any(), repository).Return(testutil.NewFakeBranchIterator(branches), nil)
+	total := numBranches*numRecords - offset
+	branchOffset := 0
+	if numRecords > 0 {
+		branchOffset = offset / numRecords
+		total = int(math.Min(float64(total/numRecords), float64(graveler.GCMaxUncommittedObjects/numRecords+1)))
+		if offset > graveler.GCMaxUncommittedObjects {
+			total += 1
+		}
+	} else {
+		total = numBranches
+	}
+
+	for i := 0; i < total; i++ {
+		test.stagingManager.EXPECT().List(gomock.Any(), gomock.Any(), gomock.Any()).Return(testutils.NewFakeValueIterator(records[i+branchOffset]), nil)
+	}
```

Comment:
😱😱😱 
I guess that this is where you spent most of your time trying to find a bug. I think it will be hard for us to maintain. I suggest to simplify it or at the very least explain what was done here.